src/systemd/sd-dns-resolver.h

```diff
@@ -0,0 +1,55 @@
+#ifndef SD_DNS_RESOLVER_H
+#define SD_DNS_RESOLVER_H
+
+#include <errno.h>
+#include <netinet/in.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#include "_sd-common.h"
+
+_SD_BEGIN_DECLARATIONS;
+
+typedef struct sd_dns_resolver sd_dns_resolver;
+
+/* https://www.iana.org/assignments/tls-extensiontype-values/tls-extensiontype-values.xhtml#alpn-protocol-ids */
+typedef enum DNSALPNFlags {
+        /* There isn't really an alpn reserved for Do53 service, but designated resolvers may or may not offer
+         * Do53 service, so we should probably have a flag to represent this capability. Unfortunately DNR
+         * does not indicate the status to us.*/
+        SD_DNS_ALPN_DO53           = 1 << 0,
+        /* SD_DNS_ALPN_HTTP_1_1,                "http/1.1" [RFC9112] */
+        SD_DNS_ALPN_HTTP_2_TLS     = 1 << 1, /* "h2"  [RFC9113] [RFC9461] */
+        /* SD_DNS_ALPN_HTTP_2_TCP,              "h2c" [RFC9113] */
+        SD_DNS_ALPN_HTTP_3         = 1 << 2, /* "h3"  [RFC9114] [RFC9461] */
+        SD_DNS_ALPN_DOT            = 1 << 3, /* "dot" [RFC7858] [RFC9461] */
+        SD_DNS_ALPN_DOQ            = 1 << 4  /* "doq" [RFC9250] [RFC9461] */
+} DNSALPNFlags;
+
+union sd_in_addr_union {
+        struct in_addr in;
+        struct in6_addr in6;
+};
+
+int sd_dns_resolver_get_priority(const sd_dns_resolver *res, uint16_t *priority);
+int sd_dns_resolver_get_adn(const sd_dns_resolver *res, const char **adn);
+int sd_dns_resolver_get_addrs(const sd_dns_resolver *res, int *family, const union sd_in_addr_union **addrs, size_t *n);
+int sd_dns_resolver_get_transports(const sd_dns_resolver *res, DNSALPNFlags *transports);
+int sd_dns_resolver_get_port(const sd_dns_resolver *res, uint16_t *port);
+int sd_dns_resolver_get_dohpath(const sd_dns_resolver *res, const char **dohpath);
+
+void sd_dns_resolver_done(sd_dns_resolver *res);
+void sd_dns_resolver_clear(sd_dns_resolver *res);
+sd_dns_resolver *sd_dns_resolver_free(sd_dns_resolver *res);
```

Comment:
so, please name this function sd_dns_resolver_unref(), even if this is internally not reference counted. As things go sooner or later we'll probably add ref counting to it, and then the function will be a misnomer.

however if you just call it sd_dns_resolver_unref() then we are prepared for a refcounting future, even if we don#t actually implement refcounting right now (i.e. no function sd_dns_resolver_ref() right now).